File: Python/Medium/AsteroidCollision.py
from typing import List
# Question: https://leetcode.com/problems/asteroid-collision/


def asteroidCollision(asteroids: List[int]) -> List[int]:

    rem = []

    for asteroid in asteroids:
        while rem and (asteroid < 0 and rem[-1] > 0):
            winner = asteroid + rem[-1]
            if winner < 0:
                rem.pop()
            elif winner > 0:
                asteroid = 0
            else:
                rem.pop()
                asteroid = 0
        if asteroid:
            rem.append(asteroid)
    return rem


if __name__ == '__main__':
    
    case1 = [5, 10, -5]
    case2 = [8, -8]
    case3 = [10, 2, -5]
    case4 = [-1, 3, 2, -3]
    case5 = [5, 10, -5, 8, -8, 2, -5]

    print(asteroidCollision(case1))
    print(asteroidCollision(case2))
    print(asteroidCollision(case3))
    print(asteroidCollision(case4))
    print(asteroidCollision(case5))

File: Java/src/Medium/AsteroidCollision.java
// https://leetcode.com/problems/asteroid-collision/?envType=problem-list-v2&envId=stack

package Medium;
import java.util.LinkedList;
import java.util.List;

public class AsteroidCollision {

    public int[] asteroidCollision(int[] asteroids) {
        
        List<Integer> rem = new LinkedList<>();

        for(int asteroid: asteroids){
            while(!rem.isEmpty() && asteroid < 0 && rem.getFirst() > 0){
                int winner = asteroid + rem.getFirst();

                if(winner < 0){
                    rem.removeFirst();
                }
                else if(winner > 0){
                    asteroid = 0;
                }
                else{
                    rem.removeFirst();
                    asteroid = 0;
                }
            }
            if(asteroid != 0){
                rem.addFirst(asteroid);
            }
        }

        return rem.stream().mapToInt(i -> i).toArray();
    }

    

    public static void main(String[] args) {
        AsteroidCollision sol = new AsteroidCollision();

        int[] c = new int[] {5,10,-5};
        System.out.println(sol.asteroidCollision(c));
        
        c = new int[] {8,-8};
        System.out.println(sol.asteroidCollision(c));
        
        c = new int[] {10,2,-5};
        System.out.println(sol.asteroidCollision(c));

    }
}
